src/tunnel.cc

```diff
@@ -1311,33 +1303,68 @@ TunnelStateData::saveError(ErrorState *error)
     savedError = error;
 }
 
-/// Starts sending the given error message to the client, leading to the
-/// eventual transaction termination. Call with savedError to send savedError.
+/// If the client expects error message, starts sending the given
+/// error message, leading to the eventual transaction termination.
+/// Otherwise initiates the transaction termination, which will happen
+/// only after all in-progress writing is complete.
 void
-TunnelStateData::sendError(ErrorState *finalError, const char *reason)
+TunnelStateData::sendErrorAndDestroy(ErrorState *err, const char *reason)
 {
     debugs(26, 3, "aborting transaction for " << reason);
 
+    // get rid of any cached error unless that is what the caller is sending
+    if (savedError != err)
+        delete savedError; // may be nil
+    savedError = nullptr;
+
+    if (finalError) {
+        debugs(26, 4, "the final error already exists: " << finalError << ", discarding new error: " << err);
+        if (finalError != err) {
+            // get rid of an error created after finalError
+            delete err; // may be nil
```

Comment:
There is a contradiction here: If `err` may be nil, then we must not delete `savedError` when `err` is nil, as `savedError` may be the error the caller wants to send. Furthermore, we should use `savedError` but we never do. There are at least two ways to address this problem:

1. Restore the `Call with savedError to send savedError` part of the API. Assert that `err` is not nil when `savedError` exists to avoid ambiguity -- does the caller wants us to ignore savedError or did the caller just forget to specify `savedError` as an explicit parameter? I do not like this approach because it is complex and relies on a dangerous runtime assertion.
2. Handle nil `err` explicitly by reusing `savedError`, if any. Calling with `savedError` is still allowed and can be useful for caller clarity sake but does not create an ambiguity.

Reply:
> There is a contradiction here: If err may be nil, then we must not delete savedError when err is nil, as savedError may be the error the caller wants to send.

The scenario you consider misses the fact that finalError exists - in this case, an _old_ savedError has been already passed to finalError but savedError could get some new value since then (which is safe to delete).

>  Furthermore, we should use savedError but we never do.

savedError is passed as an argument to this method. As far as I understood this (existing) logic - we forget about savedError existence at the moment we call sendErrorAndDestroy(). Each caller decides whether to pass saveError or create a new error on the spot.